File: LeetCode/cpp/23.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class SolutionA
{
    using valType = pair<int, int>;

public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        ListNode head(0);
        ListNode *end = &head;
        auto cmp = [](valType a, valType b) { return a.first > b.first; };
        priority_queue<valType, vector<valType>, decltype(cmp)> pq(cmp);
        for (int i = 0; i < lists.size(); ++i)
        {
            if (lists[i] != nullptr)
                pq.push(make_pair(lists[i]->val, i));
        }
        while (!pq.empty())
        {
            int ind = pq.top().second;
            pq.pop();

            end->next = lists[ind];
            end = end->next;

            lists[ind] = lists[ind]->next;
            if (lists[ind] != nullptr)
                pq.push(make_pair(lists[ind]->val, ind));
        }
        return head.next;
    }
};

class SolutionB
{
    ListNode *merge2Lists(ListNode *l1, ListNode *l2)
    {
        if (l1 == nullptr)
            return l2;
        if (l2 == nullptr)
            return l1;
        if (l1->val < l2->val)
        {
            l1->next = merge2Lists(l1->next, l2);
            return l1;
        }
        else
        {
            l2->next = merge2Lists(l1, l2->next);
            return l2;
        }
    }

public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        if (lists.empty())
            return nullptr;
        for (int step = 1; step < lists.size(); step *= 2)
        {
            for (int i = 0; i < lists.size() - step; i += step * 2)
                lists[i] = merge2Lists(lists[i], lists[i + step]);
        }
        return lists[0];
    }
};

File: lc-archived/java/23.merge-k-sorted-lists.java
/*
 * @lc app=leetcode id=23 lang=java
 *
 * [23] Merge k Sorted Lists
 *
 * https://leetcode.com/problems/merge-k-sorted-lists/description/
 *
 * algorithms
 * Hard (39.83%)
 * Likes:    6192
 * Dislikes: 329
 * Total Accepted:    784.8K
 * Total Submissions: 1.9M
 * Testcase Example:  '[[1,4,5],[1,3,4],[2,6]]'
 *
 * You are given an array of k linked-lists lists, each linked-list is sorted
 * in ascending order.
 * 
 * Merge all the linked-lists into one sorted linked-list and return it.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: lists = [[1,4,5],[1,3,4],[2,6]]
 * Output: [1,1,2,3,4,4,5,6]
 * Explanation: The linked-lists are:
 * [
 * ⁠ 1->4->5,
 * ⁠ 1->3->4,
 * ⁠ 2->6
 * ]
 * merging them into one sorted list:
 * 1->1->2->3->4->4->5->6
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: lists = []
 * Output: []
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: lists = [[]]
 * Output: []
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * k == lists.length
 * 0 <= k <= 10^4
 * 0 <= lists[i].length <= 500
 * -10^4 <= lists[i][j] <= 10^4
 * lists[i] is sorted in ascending order.
 * The sum of lists[i].length won't exceed 10^4.
 * 
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length <= 0)
            return null;

        PriorityQueue<ListNode> queue = new PriorityQueue<>((ListNode a, ListNode b) -> a.val - b.val);

        for (ListNode head : lists) {
            if (head != null)
                queue.offer(head);
        }

        ListNode dummy = new ListNode();
        ListNode cur = dummy;
        while (!queue.isEmpty()) {
            ListNode front = queue.poll();
            cur.next = front;
            cur = cur.next;
            if (front.next != null)
                queue.offer(front.next);
        }
        return dummy.next;
    }
}
// @lc code=end
